src/compiler/types/validate-build-package-json.ts

```diff
@@ -99,35 +135,87 @@ export const validateMain = (
-  const distRel = relative(config.rootDir, distAbs);
 
-  let recommendedRelPath = distRel;
-  if (customElementsOutput) {
-    const customElementsAbs = join(customElementsOutput.dir, 'index.js');
-    recommendedRelPath = relative(config.rootDir, customElementsAbs);
-  }
+  const recommendedRelPath = recommendedModulePath(config);
 
   if (!isString(currentModule)) {
-    const msg = `package.json "module" property is required when generating a distribution. It's recommended to set the "module" property to: ${recommendedRelPath}`;
+    // TODO should this change?
+    let msg = 'package.json "module" property is required when generating a distribution.';
+
+    if (recommendedRelPath !== null) {
+      msg += ` It's recommended to set the "module" property to: ${recommendedRelPath}`;
+    }
     packageJsonWarn(config, compilerCtx, buildCtx, msg, `"module"`);
-  } else if (
-    normalizePath(currentModule) !== normalizePath(recommendedRelPath) &&
-    normalizePath(currentModule) !== normalizePath(distRel)
-  ) {
+    return;
+  }
+
+  if (recommendedRelPath !== null && normalizePath(recommendedRelPath) !== normalizePath(currentModule)) {
     const msg = `package.json "module" property is set to "${currentModule}". It's recommended to set the "module" property to: ${recommendedRelPath}`;
     packageJsonWarn(config, compilerCtx, buildCtx, msg, `"module"`);
   }
 };
 
+/**
+ * Get the recommended `"module"` path for `package.json` given the output
+ * targets that a user has set on their config.
+ *
+ * @param config the user-supplied Stencil configuration
+ * @returns a recommended module path or a null value to indicate no default
+ * value is supplied
+ */
+function recommendedModulePath(config: d.Config): string | null {
+  const customElementsBundleOT = config.outputTargets.find(isOutputTargetDistCustomElementsBundle);
+  const customElementsOT = config.outputTargets.find(isOutputTargetDistCustomElements);
+  const distCollectionOT = config.outputTargets.find(isOutputTargetDistCollection);
+
+  // If we're using `dist-custom-elements` then the preferred "module" field
+  // value is `$OUTPUT_DIR/components/index.js`
+  //
+  // Additionally, the `DIST_CUSTOM_ELEMENTS` output target should override
+  // `DIST_CUSTOM_ELEMENTS_BUNDLE` and `DIST_COLLECTION` output targets if
+  // they're also set, so we return first with this one.
+  if (customElementsOT) {
+    const componentsIndexAbs = join(customElementsOT.dir, 'components', 'index.js');
+    return relative(config.rootDir, componentsIndexAbs);
+  }
+
+  if (customElementsBundleOT) {
+    const customElementsAbs = join(customElementsBundleOT.dir, 'index.js');
+    return relative(config.rootDir, customElementsAbs);
+  }
+
+  if (distCollectionOT) {
+    return relative(config.rootDir, join(distCollectionOT.dir, 'index.js'));
+  }
+
+  // if no output target for which we define a recommended output target is set
+  // we return `null`
+  return null;
+}
```

Comment:
@rwaskiewicz I made some changes here to return a proper recommended `module` path for the `DIST_COLLECTION` output target, which, in my local testing, results in the expected error message showing up for `DIST`. lmk what you think!